Allow user to define UDFs in samediff/nd4j

agibsonccc opened the pull request:
## What changes were proposed in this pull request?
Allows definition of UDFS, fixes: https://github.com/deeplearning4j/deeplearning4j/issues/9898

Users can now define custom operations while being properly integrated with samediff and serialization.
This also adds annotation scanning for registration with the samediff registered ops infrastructure.
(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [ X] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.konduit.ai/multi-project/how-to-guides/contribute/eclipse-contributors) page for details
- [X ] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [ X] Created tests for any significant new code additions.
- [ X] Relevant tests for your changes are passing.

agibsonccc committed:
Add UDFs ADR
Add new op type UDF
Update flatbuffers to reflect new changes
Add annotation scanning for udfs
Add op execution for UDFS

agibsonccc committed:
Add UDFs ADR
Add new op type UDF
Update flatbuffers to reflect new changes
Add annotation scanning for udfs
Add op execution for UDFS

agibsonccc committed:
Add UDFs ADR
Add new op type UDF
Update flatbuffers to reflect new changes
Add annotation scanning for udfs
Add op execution for UDFS
Add UDF and test for training

agibsonccc committed:
Add UDFs ADR
Add new op type UDF
Update flatbuffers to reflect new changes
Add annotation scanning for udfs
Add op execution for UDFS
Add UDF and test for training

agibsonccc committed:
Add UDFs ADR
Add new op type UDF
Update flatbuffers to reflect new changes
Add annotation scanning for udfs
Add op execution for UDFS
Add UDF and test for training
Licenses

treo commented:
I'll take a closer look later today, but in the meantime, I guess you should probably take a look at the auto generated files and the now missing license headers.

agibsonccc committed:
Add UDFs ADR
Add new op type UDF
Update flatbuffers to reflect new changes
Add annotation scanning for udfs
Add op execution for UDFS
Add UDF and test for training
Licenses
Update cuda

agibsonccc committed:
Add UDFs ADR
Add new op type UDF
Update flatbuffers to reflect new changes
Add annotation scanning for udfs
Add op execution for UDFS
Add UDF and test for training
Licenses
Update cuda

agibsonccc committed:
Add UDFs ADR
Add new op type UDF
Update flatbuffers to reflect new changes
Add annotation scanning for udfs
Add op execution for UDFS
Add UDF and test for training
Licenses
Update cuda

agibsonccc committed:
Add UDFs ADR
Add new op type UDF
Update flatbuffers to reflect new changes
Add annotation scanning for udfs
Add op execution for UDFS
Add UDF and test for training
Licenses
Update cuda

agibsonccc committed:
Add UDFs ADR
Add new op type UDF
Update flatbuffers to reflect new changes
Add annotation scanning for udfs
Add op execution for UDFS
Add UDF and test for training
Licenses
Update cuda
Remove extra

agibsonccc committed:
Add UDFs ADR
Add new op type UDF
Update flatbuffers to reflect new changes
Add annotation scanning for udfs
Add op execution for UDFS
Add UDF and test for training
Licenses
Update cuda
Remove extra generated files

agibsonccc committed:
More javadoc on doUdf
Update UDF ADR

agibsonccc committed:
Remove get/setValue

agibsonccc committed:
Add fixes for paragraphvectors deallocation
Add new event logger for more granular logs including allocations, deallocations

agibsonccc committed:
More javadoc fixes

agibsonccc committed:
More javadoc fixes

agibsonccc committed:
More javadoc fixes

agibsonccc committed:
Clean up artifacts

agibsonccc committed:
Clean up artifacts

agibsonccc committed:
Header fix

agibsonccc committed:
More fixes

agibsonccc committed:
More fixes

agibsonccc committed:
Formatting
